ros/python/py_pubsub/py_pubsub/spatial_mapper.py

```diff
@@ -0,0 +1,445 @@
+        )
+
+        self.frames_recvd = 0
+        self.prev_time = -1
+        self.meshes = {}
+
+        self.scene = trimesh.Scene()
+
+        self.poses = []
+
+        # TODO: get this from a ROS message?
+        self.image_height = 1080
+        self.image_width = 1920
+
+        log = self.get_logger()
+        #log.set_level(10)
+
+
+    def spatial_map_callback(self, msg):
+        log = self.get_logger()
+        self.frames_recvd += 1
+
+        # extract the vertices into np arrays
+        # NOTE: negating the x-component because Unity uses a left-handed
+        # coordinate system and trimesh uses a right-handed coordinate system
+        vertices = np.array([])
+        for v in msg.mesh.vertices:
+            if vertices.size == 0:
+                vertices = np.array([-v.x, v.y, v.z])
+            else:
+                vertices = np.vstack([vertices, np.array([-v.x, v.y, v.z])])
+
+        # extract the triangles into np arrays
+        triangles = np.array([])
+        for t in msg.mesh.triangles:
+            if triangles.size == 0:
+                triangles = np.array([t.vertex_indices[0],
+                                      t.vertex_indices[1],
+                                      t.vertex_indices[2]
+                                     ])
+            else:
+                triangles = np.vstack([triangles,
+                                       np.array([t.vertex_indices[0],
+                                                 t.vertex_indices[1],
+                                                 t.vertex_indices[2]
+                                                ])])
+
+        if msg.removal:
+            log.debug("Got a removal!")
+            # TODO: remove from the spatial map
+
+        mesh = trimesh.Trimesh(vertices=vertices, faces=triangles)
+        self.meshes[msg.mesh_id] = mesh
+        self.scene.add_geometry(mesh)
+
+
+    def headset_pose_callback(self, pose):
+        log = self.get_logger()
+        log.debug(f"pose stamp: {pose.header.stamp}")
+        self.poses.append(pose)
+
+
+    def detection_callback(self, detection):
+        log = self.get_logger()
+
+        if detection.num_detections == 0:
+            log.debug("No detections for this image")
+            return
+
+        # locate the headset poses for this detection using
+        # the detection source timestamp
+        world_matrix_1d = None
+        projection_matrix_1d = None
+
+        for i in range(len(self.poses)):
+            if detection.source_stamp == self.poses[i].header.stamp:
+                world_matrix_1d = self.poses[i].world_matrix
+                projection_matrix_1d = self.poses[i].projection_matrix
+
+                log.debug(f"time stamps: {self.poses[i].header.stamp} {detection.source_stamp}")
+
+                # can clear out our pose list now assuming we won't get
+                # image frame detections out of order
+                self.poses = self.poses[i:]
+                break
```

Comment:
Maybe this is OK for now (though I would suggest scanning in reverse as received detections are more likely to be associated with the more recent head poses), but this is a little expensive at `O(n)`.

We should use something that is more on the order of binary search (`O(log(n))`). Since we're doing strict equality here, a drop-in replacement would be using a binary search method using keys that are integer nanoseconds. I suppose that we are assuming we get a head pose for every frame that is encoded with the same time-stamp. If/when that assumption breaks then we need to use an approach that lets us find a most-recent-fallback (I did this recently in c++ for another project).

Reply:
I went ahead and reversed the scan and added a note about implementing something more efficient.